### PR TITLE
[1687][model] Flag for setting eval mode (on develop)

### LINKED ISSUE #1687: Introduce flag to set all frozen modules to `eval` mode
### Describe the task. It can be a feature, documentation, etc.

The model currently allows to specify modules to freeze via the `freeze_modules` parameter in the config. However, this does not mean that those parts of the model will act deterministically (e.g. due to dropout). In particular, for an ongoing experiment in the development of the diffusion model we are trying to overfit to a single encoded sample in latent space. This requires that at each iteration the frozen modules reliably output the same latent state. Hence we need a another parameter in the config that enforces not only freezing, but also sets the same modules into `eval()` mode. 

This might be useful more generally, but it should be discussed whether this is a desirable setting for other fine-tuning tasks.

Relates to #1379.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by moritzhauschulz**
## Description

This PR introduces the flag described in https://github.com/ecmwf/WeatherGenerator/issues/1687.

## Issue Number

Closes #1687 

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/model/model_interface.py [hunk 1]
```diff
@@ -70,6 +70,8 @@ def init_model_and_shard(
         if re.fullmatch(cf.freeze_modules, name) is not None:
             logger.info(f"Froze weights {name}")
             freeze_weights(module)
+        else:
```
Comment (clessig) [inline src/weathergen/model/model_interface.py]: We need to remove this. Otherwise we always have the entire model printed in the output.

File: src/weathergen/train/trainer.py [hunk 2]
```diff
@@ -405,6 +406,15 @@ def train(self, mini_epoch):
         cf = self.cf
         self.model.train()
 
+        # TODO: ablate this vs. just using no_grad
+        for name, module in self.model.named_modules():
+            name = module.name if hasattr(module, "name") else name
+            # avoid the whole model element which has name ''
+            if name == "":
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can we use 

```
re.fullmatch(cf.freeze_modules, name) is not None and name != ""
```

File: src/weathergen/train/trainer.py [hunk 3]
```diff
@@ -405,6 +406,15 @@ def train(self, mini_epoch):
         cf = self.cf
         self.model.train()
 
+        # TODO: ablate this vs. just using no_grad
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Remove comment

File: src/weathergen/train/trainer.py [hunk 4]
```diff
@@ -405,6 +406,15 @@ def train(self, mini_epoch):
...
+        for name, module in self.model.named_modules():
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: This should go to a separate function. Would it be possible to have the same function for the freezing code in model_interface and here, i.e. something like

```
def apply_fct_to_model_blocks( names, fct) : 

    for name, module in model.named_modules():
        name = module.name if hasattr(module, "name") else name
        # avoid the whole model element which has name ''
        if (re.fullmatch(cf.freeze_modules, name) is not None) and (name is not ""):
            logger.info(f"Froze weights {name}")
            fct(module)
```

and fct is then either `freeze_module()` or `module_to_eval()` (we probably find other use cases).

**Review by clessig — COMMENTED** [review]
Thanks, some smaller things to improve the code quality. Can we also add this patch:

```
diff --git a/src/weathergen/train/trainer.py b/src/weathergen/train/trainer.py
index 5ae1beab..e7b807b9 100644
--- a/src/weathergen/train/trainer.py
+++ b/src/weathergen/train/trainer.py
@@ -288,9 +288,8 @@ class Trainer(TrainerBase):
 
         # if with_fsdp then parameter count is unreliable
         if is_root():
-            if cf.with_fsdp:
-                logger.warning("Trainable parameters are inaccurate with FSDP enabled.")
-            # self.model.print_num_parameters()
+            if not cf.with_ddp:
+                self.model.print_num_parameters()
```

It's something we didn't properly clean up in #1690. But also needed to test the code, in my opinion.

File: src/weathergen/model/model_interface.py (hunk 1 above)
Comment (moritzhauschulz) [reply src/weathergen/model/model_interface.py]: I will remove this, but actually I think it is quite good to know which modules are not frozen. For example, the config prepared initially for training the diffusion engine was missing some modules that should have been frozen. I only discovered this after printing all modules. That's why I think an exhaustive list is quite useful to avoid unintended behaviour, but perhaps it should be aggregated a bit more concisely (e.g. in a list). If you agree I can open a separate issue.

File: src/weathergen/train/trainer.py (hunk 4 above)
Comment (moritzhauschulz) [reply src/weathergen/train/trainer.py]: Okay, that's a good idea.

**Review by clessig — APPROVED** [review]
Looks good. Thanks for the `apply_fct_to_blocks()` function.
